feat: filter search for commit change type

Yusin0903 opened the pull request:
## Description
Added search filter functionality to commit type selection in Conventional Commits. This allows users to quickly find commit types by typing part of the type name or description, enhancing user experience especially when working with many commit types.

Changes Made
1. Added use_search_filter: True to enable search filtering in the commit type selection.
2. Set use_jk_keys: False to prevent conflicts with search functionality (as j/k keys can be part of the search term).
3. Because the interaction questionary can not test directly so I test to verify search filter functionality and parameter works correctly.
4. Added newline at the end of file to comply with PEP 8 standards.

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `poetry all` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes

## Expected behavior
When a user types characters during commit type selection, the list of options is filtered to show only those containing the typed characters. For example, typing "fix" will show the item that contains "fix" .

## Steps to Test This Pull Request
1. Run cz commit to start the commit process
2. When the commit type selection appears, type what you want to search. eg: "fix"
3. Verify that the list filters to show the all "fix" items 
5. Try other search terms to confirm filtering works as expected

## Additional context

Lee-W commented on commitizen/cz/conventional_commits/conventional_commits.py:
```diff
@@ -209,4 +211,4 @@ def process_commit(self, commit: str) -> str:
         if m is None:
             return ""
-        return m.group(3).strip()
+        return m.group(3).strip()
```
Not sure what's system is used 🤔 but this looks like removing EOF instead

Lee-W replied:
Just fixed it for you

Lee-W commented on commitizen/cz/conventional_commits/conventional_commits.py:
```diff
@@ -43,9 +43,11 @@ class ConventionalCommitsCz(BaseCommitizen):
+                "type": "select",
                 "name": "prefix",
                 "message": "Select the type of change you are committing",
+                "use_search_filter": True,
+                "use_jk_keys": False,
```
I don't know whether we should set it as the default, probably not for v4. I would suggest making it configurable. Also, this now only works for `conventional_commits`, we would like it to work for all cz rules. Following up with this discussion, we probably could make the whole questionary behavior configurable, but that could be another PR

Yusin0903 replied:
OK, I will modify the search function to work for all cz rules. As for making it configurable, we can open another issue for that. Does that sound good?

Lee-W replied:
I would probably say we should make it configurable first to retain current behavior and let ones who want to play with it opt it

Yusin0903 replied:
Hi @Lee-W,

I have tested the search filter functionality using the configuration approach:

```toml
[[tool.commitizen.customize.questions]]
type = "list"
name = "change_type"
message = "Select the type of change you are committing"
use_search_filter = true
use_jk_keys = false
```
It works as expected, so I've:

- Removed the modifications to conventional_commits.py
- Added documentation for the feature in customization.md
- Added test cases to verify the functionality
The search filter can now be enabled through configuration without modifying the core code.

Lee-W replied:
but if we do it this way, doesn't that mean we won't be able to use it in other cz rules?

Yusin0903 replied:
My understanding is that conventional commits and JIRA were not originally intended to be customized by clients (please correct me if I'm wrong), so wouldn't it be better to just define a fixed version (like a list or select option)?

Lee-W replied:
Yes, that was the design. And I wrongly designed it 🥲 I always want to change `cz_customize` to something like `cz_empty` or `cz_no_default` and try to make everything as customizable to all rules.

Yusin0903 replied:
I wanted to ask if cz_empty and cz_no_default imply that JIRA and conventional commits can also be customized? If that's the case, should we address this in the current issue?

Lee-W replied:
> JIRA and conventional commits can also be customized? 

These two are somewhat customizable, but just not at the same level as `cz_customize`.

e.g., they have the fixed set of questions that cannot yet be customized at this moment but some of the behavior can still be changed by our config file

We don't need to do `cz_empty` in this PR. but we might want to pass these 2 options into all cz rules

Yusin0903 replied:
Not sure if this is what you meant,
But I added a question_type field in Settings, where users can choose "select" or "list".
This solution works based on my testing and fulfills the spec.

.cz.toml 
```
[tool.commitizen]
name = "cz_conventional_commits"
tag_format = "$version"
version_scheme = "semver"
version = "0.0.1"
update_changelog_on_bump = true
major_version_zero = true
question_type = "select"
```

```
def __init__(self, config: BaseConfig):
        super().__init__(config)
        self.question_type = self.config.settings.get("question_type", "list")

    def questions(self) -> Questions:
        # Default questions
        questions: Questions = [
            {
                "type": self.question_type,
                "name": "prefix",
                "message": "Select the type of change you are committing",
                ...
         if questions[0]["type"] == "select":
            questions[0].update({"use_search_filter": True, "use_jk_keys": False})
```

Lee-W replied:
Wait...

I thought we only need `use_search_filter` and `use_jk_keys` to config this kind of search on questions that is select or list?

Yusin0903 replied:
My thought is that the user only needs to set the type.
Since select type requires additional options like:

```json 
"use_search_filter": true,
"use_jk_keys": false
``` 
but list does not,
I apply these settings in code when the type is select,
so the user doesn't need to configure them manually.

Do you think config these two is a better way ?

Yusin0903 commented:
Hi @Lee-W,

I have tested the search filter functionality using the configuration approach:

[[tool.commitizen.customize.questions]]
type = "list"
name = "change_type"
message = "Select the type of change you are committing"
use_search_filter = true
use_jk_keys = false
It works as expected, so I've:

Removed the modifications to conventional_commits.py
Added documentation for the feature in customization.md
Added test cases to verify the functionality
The search filter can now be enabled through configuration without modifying the core code.

Lee-W commented:
Hey @Yusin0903 , just noticed the CI is failing. Could you please take a look? Thanks 🙂

woile commented:
We could also deprecate `cz_customize`, and just use `[tool.commitizen.customize]` to override the existining cz based on the `name`. Example: if you have set `cz_conventional_commits`, then anything in `tool.commitizen.customize` would override `cz_conventional_commits`

Lee-W commented:
> We could also deprecate `cz_customize`, and just use `[tool.commitizen.customize]` to override the existining cz based on the `name`. Example: if you have set `cz_conventional_commits`, then anything in `tool.commitizen.customize` would override `cz_conventional_commits`

Yep, this is a fantastic idea!

Lee-W reviewed (approved):
I'm good with this change. We probably want to plan the cz_customize deprecation more thoroughly
